modules/api_gateway/main.tf

```diff
@@ -0,0 +1,302 @@
+}
+
+resource "aws_api_gateway_rest_api" "rest_api" {
+  name        = var.name
+  description = var.description
+
+  endpoint_configuration {
+    types = ["EDGE"]
+  }
+
+  tags = var.tags
+}
+
+# routes
+resource "aws_api_gateway_resource" "rest_api_route_resource" {
+  for_each = local.routes
+
+  rest_api_id = aws_api_gateway_rest_api.rest_api.id
+  parent_id   = aws_api_gateway_rest_api.rest_api.root_resource_id
+  path_part   = each.key
+}
+
+resource "aws_api_gateway_method" "rest_api_route_method" {
+  for_each = local.routes
+
+  rest_api_id   = aws_api_gateway_rest_api.rest_api.id
+  resource_id   = aws_api_gateway_resource.rest_api_route_resource[each.key].id
+  http_method   = each.value["method"]
+  authorization = each.value["authorizer_key"] == "" ? "NONE" : "CUSTOM"
+  authorizer_id = (
+    each.value["authorizer_key"] == ""
+    ? null
+    : aws_api_gateway_authorizer.authorizer[each.value["authorizer_key"]].id
+  )
+}
+
+resource "aws_api_gateway_integration" "rest_api_route_integration" {
+  for_each = local.routes
+
+  rest_api_id             = aws_api_gateway_rest_api.rest_api.id
+  resource_id             = aws_api_gateway_resource.rest_api_route_resource[each.key].id
+  http_method             = aws_api_gateway_method.rest_api_route_method[each.key].http_method
+  integration_http_method = each.value["lambda_key"] != "" ? "POST" : each.value["method"]
+  type                    = each.value["type"]
+  uri = (
+    each.value["lambda_key"] != ""
+    ? replace(data.aws_lambda_function.lambda[each.value["lambda_key"]].invoke_arn, "/\\:\\d{1,3}\\/invocations/", "/invocations")
+    : each.value["proxy_url"]
+  )
+  cache_key_parameters = []
+  request_parameters   = {}
+  request_templates    = {}
+}
+
+# root route
+resource "aws_api_gateway_method" "rest_root_method" {
+  rest_api_id   = aws_api_gateway_rest_api.rest_api.id
+  resource_id   = aws_api_gateway_rest_api.rest_api.root_resource_id
+  http_method   = "ANY"
+  authorization = "CUSTOM"
+  authorizer_id = aws_api_gateway_authorizer.authorizer[var.root_route["authorizer_key"]].id
+}
+
+resource "aws_api_gateway_integration" "rest_root_integration" {
+  rest_api_id             = aws_api_gateway_rest_api.rest_api.id
+  resource_id             = aws_api_gateway_rest_api.rest_api.root_resource_id
+  http_method             = aws_api_gateway_method.rest_root_method.http_method
+  integration_http_method = "POST"
+  type                    = "AWS_PROXY"
+  uri                     = replace(data.aws_lambda_function.lambda[var.root_route["lambda_key"]].invoke_arn, "/\\:\\d{1,3}\\/invocations/", "/invocations")
+  cache_key_parameters    = []
+  request_parameters      = {}
+  request_templates       = {}
+}
+# END root route
+
+resource "aws_api_gateway_deployment" "deployment" {
+  depends_on = [
+    aws_api_gateway_integration.rest_root_integration,
+  ]
+  rest_api_id = aws_api_gateway_rest_api.rest_api.id
+
+  triggers = {
```

Comment:
This part is interesting, normally I do a `depends_on` of all aws_api_gateway_integration's and aws_api_gateway_authorizer's, which i declare statically.  However since they are dynamic i understand the need for a trigger.

I am not exactly sure how terraform is handling aws_api_gateway_deployment these days, I just know it was confusing that the aws_api_gateway_deployment didn't automatically get triggerred back in the day.

Reply:
I didn't include `depends_on` in the docs since that's a terraform global argument available for all modules. But deployments are still confusing. Here I'm trying to make them as predictable as possible.